Add equality comparisons to c10::List

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #35019 [jit] kill isSameIdentity
* #35005 [jit] kill `shallowEquals`
* #34986 [jit] define equality for IValue
* #34985 Move ivalue tests to aten
* #34892 Add equality comparison to c10::Dict
* **#34856 Add equality comparisons to c10::List**

This PR adds Python-like equality comparisons to our List type.
- `operator==` performs equality by value.
- `is` performs equality by identity.

The overall goal is that I want to define equality on `IValue` to avoid
people implementing their own broken versions. So, we should have
equality reasonably defined on all types that `IValue` can be.

@smessmer raises the concern that C++ people expect `operator==` on
reference types to test identity. I think that's a reasonable concern,
but in practice, it seems that people are defining equality functions to
do it by value anyway, just poorly. My claim is that if we just tell
people that TorchScript types behave like Python types, it will not be
super confusing.

Differential Revision: [D20483462](https://our.internmc.facebook.com/intern/diff/D20483462)

suo committed:
Add equality comparisons to c10::List

This PR adds Python-like equality comparisons to our List type.
- `operator==` performs equality by value.
- `is` performs equality by identity.

The overall goal is that I want to define equality on `IValue` to avoid
people implementing their own broken versions. So, we should have
equality reasonably defined on all types that `IValue` can be.

@smessmer raises the concern that C++ people expect `operator==` on
reference types to test identity. I think that's a reasonable concern,
but in practice, it seems that people are defining equality functions to
do it by value anyway, just poorly. My claim is that if we just tell
people that TorchScript types behave like Python types, it will not be
super confusing.

[ghstack-poisoned]

suo committed:
Update on "Add equality comparisons to c10::List"


This PR adds Python-like equality comparisons to our List type.
- `operator==` performs equality by value.
- `is` performs equality by identity.

The overall goal is that I want to define equality on `IValue` to avoid
people implementing their own broken versions. So, we should have
equality reasonably defined on all types that `IValue` can be.

@smessmer raises the concern that C++ people expect `operator==` on
reference types to test identity. I think that's a reasonable concern,
but in practice, it seems that people are defining equality functions to
do it by value anyway, just poorly. My claim is that if we just tell
people that TorchScript types behave like Python types, it will not be
super confusing.

Differential Revision: [D20483462](https://our.internmc.facebook.com/intern/diff/D20483462)

[ghstack-poisoned]

suo committed:
Update on "Add equality comparisons to c10::List"


This PR adds Python-like equality comparisons to our List type.
- `operator==` performs equality by value.
- `is` performs equality by identity.

The overall goal is that I want to define equality on `IValue` to avoid
people implementing their own broken versions. So, we should have
equality reasonably defined on all types that `IValue` can be.

@smessmer raises the concern that C++ people expect `operator==` on
reference types to test identity. I think that's a reasonable concern,
but in practice, it seems that people are defining equality functions to
do it by value anyway, just poorly. My claim is that if we just tell
people that TorchScript types behave like Python types, it will not be
super confusing.

Differential Revision: [D20483462](https://our.internmc.facebook.com/intern/diff/D20483462)

[ghstack-poisoned]

bhosmer commented on aten/src/ATen/core/List.h:
```diff
@@ -392,12 +391,21 @@ class List final {
+   * Identity comparison. Returns true if and only if `rhs` represents the same
+   * List object as `this`.
+   */
+  bool is(const List<T>& rhs) const;
```
Just a style thing, but noting that here you're giving the type argument explicitly for the param, but not in `Dict::is`.

...also (for both List and Dict), I'm assuming wanted `is` defined only within the specialization, right?

suo committed:
Update on "Add equality comparisons to c10::List"


This PR adds Python-like equality comparisons to our List type.
- `operator==` performs equality by value.
- `is` performs equality by identity.

The overall goal is that I want to define equality on `IValue` to avoid
people implementing their own broken versions. So, we should have
equality reasonably defined on all types that `IValue` can be.

@smessmer raises the concern that C++ people expect `operator==` on
reference types to test identity. I think that's a reasonable concern,
but in practice, it seems that people are defining equality functions to
do it by value anyway, just poorly. My claim is that if we just tell
people that TorchScript types behave like Python types, it will not be
super confusing.

Differential Revision: [D20483462](https://our.internmc.facebook.com/intern/diff/D20483462)

[ghstack-poisoned]

suo committed:
Update on "Add equality comparisons to c10::List"


This PR adds Python-like equality comparisons to our List type.
- `operator==` performs equality by value.
- `is` performs equality by identity.

The overall goal is that I want to define equality on `IValue` to avoid
people implementing their own broken versions. So, we should have
equality reasonably defined on all types that `IValue` can be.

@smessmer raises the concern that C++ people expect `operator==` on
reference types to test identity. I think that's a reasonable concern,
but in practice, it seems that people are defining equality functions to
do it by value anyway, just poorly. My claim is that if we just tell
people that TorchScript types behave like Python types, it will not be
super confusing.

Differential Revision: [D20483462](https://our.internmc.facebook.com/intern/diff/D20483462)

[ghstack-poisoned]
